### PR TITLE
Directions optimization demo

### PR DESCRIPTION
**Pull request by electrostat**
- route optimization demo that is comparable to web demo

### CODE REVIEW

**Comment by langsmith** [comment]
_Note: Fixes #281_

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/MainActivity.java [hunk 1]
```diff
@@ -506,6 +507,12 @@ private void listItems(int id) {
           R.string.activity_mas_directions_url
         ));
         exampleItemModel.add(new ExampleItemModel(
+          R.string.activity_mas_optimization_title,
+          R.string.activity_mas_optimization_description,
+          new Intent(MainActivity.this, OptimizationActivity.class),
+          R.string.activity_mas_directions_url
```
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/MainActivity.java]: Please change to `activity_mas_optimization_url` for continuity with `activity_mas_optimization_title` and `activity_mas_optimization_description` strings

Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/MainActivity.java]: Nevermind. I'll change it

**Comment by langsmith** [comment]
Also, no worries if you see the jibberish in the nav drawer (screenshot below). It's from me forgetting to delete the menu item [when I removed all of the location examples](https://github.com/mapbox/mapbox-android-demo/pull/424).

[I've made a new pr to fix it](https://github.com/mapbox/mapbox-android-demo/pull/434).

device-2017-08-02-150820

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+        System.out.println(call.request().url().toString());
+
+        // You can get the generic HTTP info about the response
+        Log.d(TAG, "Response code: " + response.body().getTrips());
+        Log.d(TAG, "response: " + call.request().url().toString());
+
+        if (response.body() == null) {
+          Log.e(TAG, "No routes found, make sure you set the right user and access token.");
+          return;
+        } else if (response.body().getTrips().size() < 1) {
```
Comment (Guardiola31337) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: In order to make the statement more semantic, why not using `response.body().getTrips().isEmpty()` instead?

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+          Log.e(TAG, "No routes found" + response.body().getTrips().size());
+          return;
+        }
+
+        // Print some info about the route
+        optimizedRoute = response.body().getTrips().get(0);
+
+        // Draw the route on the map
```
Comment (Guardiola31337) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: Useless comment.
> The best comment is a good name for a method or class.

And in this case the name of the method (`drawOptimized`) is clear.

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+      }
+
+      @Override
+      public void onFailure(Call<OptimizedTripsResponse> call, Throwable throwable) {
+        Log.e(TAG, "Error: " + throwable.getMessage());
+      }
+    });
+  }
+
+  private void drawOptimized(DirectionsRoute route) {
```
Comment (Guardiola31337) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: Method a bit long. What about extracting commented blocks of code into well-named methods?

Comment (Guardiola31337) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: Method a bit long. What about extracting the blocks of code followed by a comment into well-named methods?

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+    for (int i = 0; i < coordinates.size(); i++) {
+      points[i] = new LatLng(
+        coordinates.get(i).getLatitude(),
+        coordinates.get(i).getLongitude());
+    }
+
+    // Draw Points on MapView
+    optimizedPolyline = map.addPolyline(new PolylineOptions()
+      .add(points)
+      .color(Color.parseColor("#23D2BE"))
```
Comment (Guardiola31337) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: Magic number. What color is that?

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+    mapView.onCreate(savedInstanceState);
+    mapView.getMapAsync(new OnMapReadyCallback() {
+      @Override
+      public void onMapReady(MapboxMap mapboxMap) {
+        map = mapboxMap;
+
+        // Add origin and destination to the map
+        mapboxMap.addMarker(new MarkerOptions()
+          .position(new LatLng(origin.getLatitude(), origin.getLongitude()))
+          .title("1"));
```
Comment (Guardiola31337) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: `"1"` isn't a great title. What about `"origin"`?

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+          @Override
+          public void onMapClick(@NonNull LatLng point) {
+
+            // Optimization API is limited to 12 coordinate sets
+            if (stops.size() == 12) {
+              Toast.makeText(OptimizationActivity.this, "Only 12 stops allowed on route", Toast.LENGTH_LONG).show();
+            } else {
+              map.addMarker(new MarkerOptions()
+                .position(new LatLng(point.getLatitude(), point.getLongitude()))
+                .title("1"));
```
Comment (Guardiola31337) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: Same here. How does `"destination"` sound?

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java (hunk 6 above)
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: Fixed

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java (hunk 7 above)
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: Fixed

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+              getOptimized(stops);
+            }
+          }
+        });
+      }
+    });
+  }
+
+  private void getOptimized(List<Position> coordinates) {
+
```
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: Removed

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+    optimizedClient = new MapboxOptimizedTrips.Builder()
+      .setSource("first")
```
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: Fixed

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+      .setCoordinates(coordinates)
+      .setOverview(DirectionsCriteria.OVERVIEW_FULL)
+      .setProfile(DirectionsCriteria.PROFILE_DRIVING)
+      .setAccessToken(Mapbox.getAccessToken())
+      .build();
+
+    optimizedClient.enqueueCall(new Callback<OptimizedTripsResponse>() {
+      @Override
+      public void onResponse(Call<OptimizedTripsResponse> call, Response<OptimizedTripsResponse> response) {
+        System.out.println(call.request().url().toString());
```
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: Fixed and now using `Log`

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java (hunk 3 above)
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: Removed

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java (hunk 5 above)
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: Changed to `.color(Color.parseColor(TEAL_COLOR))` with `private static final String TEAL_COLOR = "23D2BE";` at the top

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java (hunk 4 above)
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: Fixed

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java (hunk 2 above)
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: Fixed

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+
+        // You can get the generic HTTP info about the response
+        Log.d(TAG, "Response code: " + response.body().getTrips());
+        Log.d(TAG, "response: " + call.request().url().toString());
+
+        if (response.body() == null) {
```
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: Fixed

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+      .setAccessToken(Mapbox.getAccessToken())
+      .build();
+
+    optimizedClient.enqueueCall(new Callback<OptimizedTripsResponse>() {
+      @Override
+      public void onResponse(Call<OptimizedTripsResponse> call, Response<OptimizedTripsResponse> response) {
+        System.out.println(call.request().url().toString());
+
+        // You can get the generic HTTP info about the response
+        Log.d(TAG, "Response code: " + response.body().getTrips());
```
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: Removed the log line all together 'cause it's not needed

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+        Log.d(TAG, "response: " + call.request().url().toString());
```
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: removed log line 'cause not needed

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+    // object or in the same activity which contains the mapview.
+    Mapbox.getInstance(this, getString(R.string.access_token));
+
+    // This contains the MapView in XML and needs to be called after the access token is configured.
+    setContentView(R.layout.activity_mas_optimization);
+
+    //Set up stop list
+    stops = new ArrayList<Position>();
+
+    // Set First Stop
```
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: Done. `initializeListOfStops()` is now run in `onCreate()`

```java
private void initializeListOfStops() {
    //Set up stop list
    stops = new ArrayList<>();

    // Set first stop
    final Position origin = Position.fromCoordinates(-122.408818, 37.784015);
    stops.add(origin);
  }
```

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java (hunk 4 above)
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: Done

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+        mapboxMap.addMarker(new MarkerOptions()
+          .position(new LatLng(origin.getLatitude(), origin.getLongitude()))
+          .title("1"));
+
+        map.setOnMapClickListener(new MapboxMap.OnMapClickListener() {
+          @Override
+          public void onMapClick(@NonNull LatLng point) {
+
+            // Optimization API is limited to 12 coordinate sets
+            if (stops.size() == 12) {
```
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: Done

**Comment by langsmith** [comment]
@Guardiola31337 , I've addressed/fixed all of your review comments and example's working fine.

ezgif com-resize

**Comment by langsmith** [comment]
Ready for review @zugaldia  

ezgif com-resize

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java [hunk 16]
```diff
@@ -0,0 +1,238 @@
...
+import java.util.List;
+
+import retrofit2.Call;
+import retrofit2.Callback;
+import retrofit2.Response;
+
+import static com.mapbox.services.Constants.PRECISION_6;
+
+/**
+ * Use Mapbox Android Services to request and compare normal directions with time optimized directions
```
Comment (zugaldia) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: nit: missing final dot.

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+    map.addMarker(new MarkerOptions()
+      .position(new LatLng(point.getLatitude(), point.getLongitude()))
+      .title(getString(R.string.destination)));
+  }
+
+  private void addPointToStopsList(LatLng point) {
+    stops.add(Position.fromCoordinates(point.getLongitude(), point.getLatitude()));
+  }
+
+
```
Comment (zugaldia) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: nit: extra line

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java [hunk 18]
```diff
@@ -0,0 +1,238 @@
...
+  private void initializeListOfStops() {
+    //Set up stop list
```
Comment (zugaldia) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: nit: missing space after `//`

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java [hunk 19]
```diff
@@ -0,0 +1,238 @@
...
+      .setCoordinates(coordinates)
+      .setOverview(DirectionsCriteria.OVERVIEW_FULL)
+      .setProfile(DirectionsCriteria.PROFILE_DRIVING)
+      .setAccessToken(Mapbox.getAccessToken())
+      .build();
+
+    optimizedClient.enqueueCall(new Callback<OptimizedTripsResponse>() {
+      @Override
+      public void onResponse(Call<OptimizedTripsResponse> call, Response<OptimizedTripsResponse> response) {
+
```
Comment (zugaldia) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: nit: extra line

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java [hunk 20]
```diff
@@ -0,0 +1,238 @@
...
+              + response.body().getTrips().size());
+            Toast.makeText(OptimizationActivity.this, R.string.successful_but_no_routes,
+              Toast.LENGTH_SHORT).show();
+            return;
+          }
+        }
+
+        // Get most optimized route from API response
+        optimizedRoute = response.body().getTrips().get(0);
+
```
Comment (zugaldia) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: nit: extra line

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java [hunk 21]
```diff
@@ -0,0 +1,238 @@
...
+  }
+
+  private void drawOptimizedRoute(DirectionsRoute route) {
+    // Remove old polyline
+    if (optimizedPolyline != null) {
+      map.removePolyline(optimizedPolyline);
+    }
+
+    LatLng[] pointsToDraw = convertLineStringToLatLng(route);
+
```
Comment (zugaldia) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: nit: extra line, comment could go before `LatLng[] pointsToDraw...`

File: MapboxAndroidDemo/src/main/res/values/strings.xml [hunk 22]
```diff
@@ -412,8 +415,16 @@
     <string name="rv_card_bed_info">This location has %1$d beds</string>
     <string name="toast_instruction">Click on cards below</string>
 
+    <!-- Optimization API activity -->
+    <string name="origin">Origin</string>
+    <string name="destination">Destination</string>
+    <string name="only_twelve_stops_allowed">Only 12 stops allowed on route</string>
+    <string name="click_instructions">Tap on the map to select a maximum of 12 locations</string>
+    <string name="successful_but_no_routes">No routes found for that location</string>
+    <string name="no_success">No routes found, make sure you set the right user and access token.</string>
```
Comment (zugaldia) [inline MapboxAndroidDemo/src/main/res/values/strings.xml]: nit: remove final period.

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java [hunk 23]
```diff
@@ -0,0 +1,238 @@
...
+    super.onCreate(savedInstanceState);
+
+    // Mapbox access token is configured here. This needs to be called either in your application
+    // object or in the same activity which contains the mapview.
+    Mapbox.getInstance(this, getString(R.string.access_token));
+
+    // This contains the MapView in XML and needs to be called after the access token is configured.
+    setContentView(R.layout.activity_mas_optimization);
+
+    initializeListOfStops();
```
Comment (zugaldia) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: I'd add a comment about what this method will do.

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java (hunk 16 above)
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: Added

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java (hunk 23 above)
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: Added

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java (hunk 17 above)
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: Removed

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java (hunk 18 above)
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: Added

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java (hunk 19 above)
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: Removed

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java (hunk 20 above)
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: Removed

File: MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java (hunk 21 above)
Comment (langsmith) [inline MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/mas/OptimizationActivity.java]: Changed

File: MapboxAndroidDemo/src/main/res/values/strings.xml (hunk 22 above)
Comment (langsmith) [inline MapboxAndroidDemo/src/main/res/values/strings.xml]: Removed

**Pull request merged**